02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,225 @@
+import { SampledValueType } from '@citrineos/base/dist/ocpp/model/types/MeterValuesRequest';
+
+/**
+ * Util to process and validate signed meter values.
+ */
+export class SignedMeterValuesUtil {
+  private readonly _fileAccess: IFileAccess;
+  private readonly _logger: Logger<ILogObj>;
+  private readonly _deviceModelRepository: IDeviceModelRepository;
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  /**
+   * Checks the validity of a meter value.
+   *
+   * If a meter value is unsigned, it is valid.
+   *
+   * If a meter value is signed, it is valid if:
+   * - SignedMeterValuesConfig is configured
+   * AND
+   * - The incoming signed meter value's signing method matches the configured signing method
+   * AND
+   * - The incoming signed meter value's public key is empty but there is a public key stored for that charging station
+   * OR
+   * - The incoming signed meter value's public key isn't empty and it matches the configured public key
+   *
+   * @param stationId - The charging station the meter values belong to
+   * @param meterValues - The list of meter values
+   */
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const publicKeyFrequencyVariableAttribute =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    const shouldPublicKeyBeUsedForValidation =
+      publicKeyFrequencyVariableAttribute.length > 0 &&
+      publicKeyFrequencyVariableAttribute[0].value !== 'Never';
+
+    if (shouldPublicKeyBeUsedForValidation) {
```

Comment:
Not sure about this.

If charger is set to never send public key, then I'd think that we should still to validation but should just use the CSMS configured public key.

cc: @thanaParis

Reply:
Honestly, I'm not sure we really need to fetch this variable attribute in the first place.

We can just do a check on this `publicKey` field of the SignedMeterValue to check if it is non-empty.

If it is present, then check to see if it matches the public key that the CSMS has configured, and proceed with validation. If it is not present, then just do validation.

Reply:
rolling around this logic in my head, i see what you mean. i think i was using this initially to choose whether or not to do the public key processing. but if a signed meter value exists in general, we don't really care if a public key is in the message or not bc we already have logic that can check the DB for the public key instead, and we'd always want to do the validation if we have a signed meter value 🤔 

let me simplify the logic 👍